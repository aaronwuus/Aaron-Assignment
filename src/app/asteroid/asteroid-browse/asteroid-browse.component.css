.desc {
    display: block;
    line-height: 38px;
    font-stretch: bold;
    font-size: 16px;
    color: blue;
}
.thead {
    color: navy;
    font-stretch: bold;
    font-size: 16px;
    vertical-align: middle;
}
.isHazard {
    background-color:orangered;
    font-stretch: bold;
    color: blue;
}
table td {
    position: relative;
    text-align: center;
    padding: 0;
    margin: 0;
}
input {
    text-align: center;
    border: 0px;
    background-color: beige;
}
textarea:disabled, input:disabled {
    text-align: center;
    background-color: transparent;
    border: 0px;
}