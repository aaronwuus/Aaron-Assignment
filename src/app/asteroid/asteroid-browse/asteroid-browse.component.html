<div class="row">
    <div class="col-8 text-center" >
        <span class="desc">Retrieve Near Earth Asteroid Data Through NASA API ! Click Edit Icon To Toggle Editable Function</span>
    </div>
<div class= "col-4">
    <app-pagination [current]="getPageData().current" [max]="getPageData().totalPage" [prev]="getPageData().prev" [next]="getPageData().next" (changePage)="pageChanged($event)"></app-pagination>
</div>
<div class="container-fluid">
    <div class="col col-sm-2">
    </div>
    <div class="col-sm-auto">
    <table class="table table-striped table-bordered table-hover">
    <thead class="thead">
        <tr>
        <th scope="col">Reference<br/>ID</th>
        <th scope="col">Name</th>
        <th scope="col">Designation</th>
        <th scope="col">Magnitude</th>
        <th scope="col">Diameter<br/>(feet)</th>
        <th scope="col">Potentially<br/>Hazardous</th>
        </tr>
    </thead>
    <tbody>
        <tr class="text-center" *ngFor="let object of responeseData.near_earth_objects; let i = index" (appInvoke)="iterateEditStatusTable(i,7)">
            <td scope="col">
                <input [attr.disabled]="editStatusArray[i][0] ? '' : null" 
                        [(ngModel)]='object.id' 
                        name="id"/>
                <span style="float:right;" (click)="editadble(i,0)" >
                    <i class="fas fa-edit"></i>
                </span>
            </td>
            <td scope="col">
                <input [attr.disabled]="editStatusArray[i][1] ? '' : null" 
                        [(ngModel)]='object.name' 
                        name="name"/>
                <span style="float:right;" (click)="editadble(i,1)">
                    <i class="fas fa-edit"></i>
                </span>
            </td>
            <td scope="col">
                <input [attr.disabled]="editStatusArray[i][2] ? '' : null" 
                        [(ngModel)]='object.designation' 
                        name="designation"/> 
                <span style="float:right;" (click)="editadble(i,2)">
                    <i class="fas fa-edit"></i>
                </span>
            </td>
            <td scope="col">
                <input [attr.disabled]="editStatusArray[i][3] ? '' : null" 
                        [(ngModel)]='object.absolute_magnitude_h' 
                        name="magnitude"/> 
                <span style="float:right;" (click)="editadble(i,3)">
                    <i class="fas fa-edit"></i>
                </span>
            </td>
            <td scope="col">
                <input [attr.disabled]="editStatusArray[i][4] ? '' : null" 
                        [ngModel]="object.estimated_diameter.feet.estimated_diameter_min | number:'.2-2'" 
                        (ngModelChange)="object.estimated_diameter.feet.estimated_diameter_min=$event"
                        name="diameter_min"/>
                <span>-</span>
                <input [attr.disabled]="editStatusArray[i][4] ? '' : null" 
                        [ngModel]="object.estimated_diameter.feet.estimated_diameter_max | number:'.2-2'" 
                        (ngModelChange)="object.estimated_diameter.feet.estimated_diameter_max=$event"
                        name="diameter_max"/> 
                <span style="float:right;" (click)="editadble(i,4)">
                    <i class="fas fa-edit"></i>
                </span>
            </td>
            <td scope="col">
                <input [attr.disabled]="editStatusArray[i][5] ? '' : null"  
                        [ngModel]='object.is_potentially_hazardous_asteroid | boolean' 
                        (ngModelChange)="object.is_potentially_hazardous_asteroid"/> 
                <span style="float:right;" (click)="editadble(i,5)">
                    <i class="fas fa-edit"></i>
                </span>
            </td>
        </tr>
    </tbody>
    </table>
    </div>
    <div class="col col-sm-2">
    </div>
</div>