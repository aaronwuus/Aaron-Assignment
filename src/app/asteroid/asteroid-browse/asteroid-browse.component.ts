import { Subscription } from 'rxjs';
import { NavbarService } from './../../services/navbar.service';
import { Pagination } from './../../models/Pagination';
import { Links } from './../../models/Links';
import { environment } from './../../../environments/environment';
import { Page } from './../../models/Page';
import { NeoBrowseService } from './../../services/neo-browse.service';
import { ResponseData } from './../../models/ResponseData';
import { Component, OnInit, OnDestroy } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { NearEarthObject } from 'src/app/models/NearEarthObject';

@Component({
  selector: 'app-asteroid-browse',
  templateUrl: './asteroid-browse.component.html',
  styleUrls: ['./asteroid-browse.component.css']
})
export class AsteroidBrowseComponent implements OnInit, OnDestroy {

  responeseData: ResponseData;
  apiKey: string;
  links: Links;
  page: Page;
  queryParams: string;
  nearEarthObject: NearEarthObject[];
  pagination: Pagination;
  isHazard: boolean;
  readonly = false;
  editStatusArray: boolean[][] = [];
  private initSubscription: Subscription;
  private onChangeSubscription: Subscription;
  constructor(private neoBrowseService: NeoBrowseService,
              private activatedRoute: ActivatedRoute,
              private navbarService: NavbarService, ) {
    this.page = { size: 20, total_elements: 0, total_pages: 0, number: 0 } as Page;
    this.apiKey = '';
    this.links = { next: '', prev: '', self: '' };
    this.nearEarthObject = [ ] as NearEarthObject[];
    this.responeseData = { links: this.links, page: this.page, near_earth_objects: this.nearEarthObject } as ResponseData;
    this.pagination = { prev: 0, current: 0, next: 0, totalElements: 0, totalPage: 0, pageSize: 0, path: '' } as Pagination;
    this.isHazard = false;
    this.editStatusArray[0] = [];
    for (let  i = 0; i < 20 ; i++ ) {
        this.editStatusArray[i] = [true, true, true, true, true, true, true];
        this.editStatusArray.push([]);
    }
   }

  ngOnInit() {
    this.activatedRoute.paramMap.subscribe(par => this.page.number = + par.get('index') );
    this.queryParams = `?page=${this.page.number - 1}&size=${this.page.size}&api_key=${environment.apikey}`;
    this.initSubscription = this.neoBrowseService.neoBrowse(environment.asteroidBrowseBaseUrl + this.queryParams)
      .subscribe(response => {
        this.responeseData = response as ResponseData;
      });
    this.neoBrowseService.navPage.next(this.page.number);
    this.navbarService.show();
  }

  getPageData(): Pagination {
    this.pagination.current = this.page.number;
    this.pagination.next = this.page.number >= this.responeseData.page.total_pages
                                          ? this.responeseData.page.total_pages
                                          : this.page.number + 1;
    this.pagination.prev = this.page.number <= 1 ? 1 : this.page.number - 1;
    this.pagination.pageSize = this.responeseData.page.size;
    return this.pagination;
  }

  pageChanged(event): Pagination {
    this.page.number = event;
    this.queryParams = `?page=${this.page.number - 1}&size=${this.page.size}&api_key=${environment.apikey}`;
    this.onChangeSubscription = this.neoBrowseService.neoBrowse(environment.asteroidBrowseBaseUrl + this.queryParams)
      .subscribe(response => {
        this.responeseData = response as ResponseData;
      });
    this.neoBrowseService.navPage.next(this.page.number);
    return this.getPageData();
  }

  editadble(index: number, column: number) {
    this.editStatusArray[index][column] = !this.editStatusArray[index][column] ;
  }

  iterateEditStatusTable(index: number, column: number) {
    for (let  i = 0; i < column ; i++ ) {
      this.editStatusArray[index][i] = false;
    }
  }
  ngOnDestroy(): void {
    if (this.initSubscription !== undefined) {
    this.initSubscription.unsubscribe();
    }
    if (this.onChangeSubscription !== undefined) {
    this.onChangeSubscription.unsubscribe();
    }
  }
}
