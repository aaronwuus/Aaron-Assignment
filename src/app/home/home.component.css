section {
    height: 100vh; 
    background: rgb(245, 250, 185);
    background-image: 
        linear-gradient(45deg, rgba(5, 190, 240, 0.4) 25%, transparent 0, transparent 75%, rgba(5, 190, 240, 0.4) 0),
        linear-gradient(45deg, rgba(5, 190, 240, 0.4) 25%, transparent 0, transparent 75%, rgba(5, 190, 240, 0.4) 0);
    background-position: 0 0, 85px 85px;
    background-size: 170px 170px;
}
.definition {
    border: 5px solid rgba(250, 150, 150, 0.7);
    border-radius: 50%;
    display: inline-block;
    position: relative;
    vertical-align: middle;
    margin: 0;
    width: 200px;
    height: 200px;
    text-align: center;
}
.definition_teaser .definition:before {
    content: '';
    display: inline-block;
    height: 30%;
    vertical-align: middle;
    margin-top: .5em;
    margin-left: 0;
    overflow: hidden;
    position: relative;
}
h4 {
    margin-left: 10px;
}
.grid_layout {
    display: flex;
    align-items: center;
    height: 220px;
    position: fixed;;
    top: 30%;
    left: 30%;
} 
.active {
    color: rgb(25, 55, 25);
    text-decoration: none;
}
a {
    height: 100%;
    transition:0.3s;
    color: rgb(25, 55, 25);
    text-decoration: none;
    padding-left: 50px;
    padding-right: 50px;
} 