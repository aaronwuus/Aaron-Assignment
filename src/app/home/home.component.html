<section class="definition_teaser">
    <div class="grid_layout">
        <span class="definition_pairs">
            <a [routerLink]="[ '/browse', '1' ]" routerLinkActive="active">
                <div class="definition">
                <h4 class="title">Problem One<br/>Click Me</h4>
                </div>
            </a>
            <a [routerLink]="[ '/widget' ]" routerLinkActive="active">
                <div class="definition">
                <h4 class="title">Problem Two<br/>Click Me</h4>
                </div>
            </a>
        </span>
    </div>
</section>

