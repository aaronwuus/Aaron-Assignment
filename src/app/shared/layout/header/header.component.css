.navbar-custom {
    background-color: #99CCCC;
    font-size: 25px;
    font-weight: normal;
    color: #FCFCFC;
    padding: 0px;
    height      : 75px;
    line-height : 75px;
}
.title {
    color: #FCFCFC;
}
.left-nav {
    margin-right: 50px;

}
.right-nav {
    margin-left: 50px;

}
.active {
    background: rgba(65, 75, 75, 0.397) ;
    text-decoration: none;
}
a {
    display: block;
    height: 100%;
    transition:0.3s;
    color: #FCFCFC;
    text-decoration: none;
    padding-left: 5px;
    padding-right: 5px;
}
