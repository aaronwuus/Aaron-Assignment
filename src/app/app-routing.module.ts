import { WidgetComponent } from './widget/widget.component';
import { HomeComponent } from './home/home.component';
import { PageNotFoundComponent } from './shared/page-not-found/page-not-found.component';
import { AsteroidBrowseComponent } from './asteroid/asteroid-browse/asteroid-browse.component';
import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

const routes: Routes = [
  { path: '', component: HomeComponent },
  { path: 'browse/:index', component: AsteroidBrowseComponent },
  { path: 'widget', component: WidgetComponent },
  { path: '**', component: PageNotFoundComponent }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
