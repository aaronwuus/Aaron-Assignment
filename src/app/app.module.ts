import { BooleanPipe } from './shared/pipes/boolean.pipe';
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';
import { AsteroidBrowseComponent } from './asteroid/asteroid-browse/asteroid-browse.component';
import { PageNotFoundComponent } from './shared/page-not-found/page-not-found.component';
import { HttpClientModule } from '@angular/common/http';
import { TrimPipe } from './shared/pipes/trim.pipe';
import { HeaderComponent } from './shared/layout/header/header.component';
import { HomeComponent } from './home/home.component';
import { PaginationComponent } from './shared/pagination/pagination.component';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { FormsModule } from '@angular/forms';
import { InvokeDirective } from './shared/directives/invoke.directive';
import { WidgetOneComponent } from './widget/widget-one/widget-one.component';
import { WidgetTwoComponent } from './widget/widget-two/widget-two.component';
import { WidgetComponent } from './widget/widget.component';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ProductListComponent } from './product-list/product-list.component';
import { TopBarComponent } from './top-bar/top-bar.component';

@NgModule({
  declarations: [
    AppComponent,
    AsteroidBrowseComponent,
    PageNotFoundComponent,
    TrimPipe,
    HeaderComponent,
    HomeComponent,
    PaginationComponent,
    BooleanPipe,
    InvokeDirective,
    WidgetOneComponent,
    WidgetTwoComponent,
    WidgetComponent,
    ProductListComponent,
    TopBarComponent
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    HttpClientModule,
    NgbModule,
    FormsModule
  ],
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
