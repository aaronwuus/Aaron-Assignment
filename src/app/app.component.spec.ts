import { AsteroidBrowseComponent } from './asteroid/asteroid-browse/asteroid-browse.component';
import { HeaderComponent } from './shared/layout/header/header.component';
import { PageNotFoundComponent } from './shared/page-not-found/page-not-found.component';
import { PaginationComponent } from './shared/pagination/pagination.component';
import { WidgetTwoComponent } from './widget/widget-two/widget-two.component';
import { WidgetOneComponent } from './widget/widget-one/widget-one.component';
import { WidgetComponent } from './widget/widget.component';
import { HomeComponent } from './home/home.component';
import { AppRoutingModule } from './app-routing.module';
import { BooleanPipe } from './shared/pipes/boolean.pipe';
import { WidgetService } from './services/widget.service';
import { NeoBrowseService } from './services/neo-browse.service';
import { NavbarService } from './services/navbar.service';
import { ApiService } from './services/api.service';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { TestBed, async } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { AppComponent } from './app.component';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { APP_BASE_HREF } from '@angular/common';

describe('AppComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        HttpClientModule,
        FormsModule,
        NgbModule,
        AppRoutingModule
      ],
      declarations: [
        AppComponent,
        BooleanPipe,
        HomeComponent,
        WidgetComponent,
        WidgetOneComponent,
        WidgetTwoComponent,
        PaginationComponent,
        PageNotFoundComponent,
        HeaderComponent,
        AsteroidBrowseComponent
      ],
      providers: [ApiService, NavbarService, NeoBrowseService, WidgetService, {provide: APP_BASE_HREF, useValue: '/'}],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();
  }));

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app).toBeTruthy();
  });

  it(`should have as title 'assignment'`, () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app.title).toEqual('assignment');
  });

  it('should render title in a h1 tag', () => {
    const fixture = TestBed.createComponent(AppComponent);
    // fixture.detectChanges();
    // const compiled = fixture.debugElement.nativeElement;
    // expect(compiled.querySelector('h1').textContent).toContain('Welcome to assignment!');
    expect(fixture).toMatchSnapshot();
  });

});
