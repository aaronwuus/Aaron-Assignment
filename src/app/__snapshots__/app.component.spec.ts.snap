// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should render title in a h1 tag 1`] = `
<app-root
  title={[Function String]}
>
  <app-header>
    
  </app-header><div
    class="container-fluid"
    style="padding-left: 0px;padding-right: 0px;"
  >
    <router-outlet />
  </div>
</app-root>
`;
